examples/how-to/perform-http-requests/tests/http_requests.rs

```diff
@@ -0,0 +1,248 @@
+    };
+
+    let bytes = byte_list
+        .iter()
+        .map(|value| {
+            value
+                .as_i64()
+                .ok_or(())
+                .and_then(|integer| integer.try_into().map_err(|_| ()))
+        })
+        .collect::<Result<Vec<u8>, _>>()
+        .unwrap_or_else(|()| {
+            panic!("Expected a list of bytes representing the response body, got {byte_list:#?}")
+        });
+
+    assert_eq!(bytes, HTTP_RESPONSE_BODY.as_bytes());
+
+    Ok(())
+}
+
+/// Tests if service query can't perform HTTP requests to hosts that aren't allowed.
+#[test_log::test(tokio::test)]
+#[should_panic(expected = "UnauthorizedHttpRequest")]
+async fn service_query_cant_send_http_request_to_unauthorized_host() {
+    let url = "http://localhost/".to_owned();
+
+    let (_validator, application_id, chain) =
+        TestValidator::with_current_application::<Abi, _, _>(url, ()).await;
+
+    chain
+        .graphql_query(application_id, "query { performHttpRequest }")
+        .await;
+}
+
+/// Tests if the service sends a valid HTTP response to the contract.
+#[test_log::test(tokio::test)]
+async fn service_sends_valid_http_response_to_contract() -> anyhow::Result<()> {
+    const HTTP_RESPONSE_BODY: &str = "Hello, world!";
+
+    let http_server =
+        HttpServer::start(Router::new().route("/", get(|| async { HTTP_RESPONSE_BODY }))).await?;
+    let port = http_server.port();
+    let url = format!("http://localhost:{port}/");
+
+    let (validator, application_id, chain) =
+        TestValidator::with_current_application::<Abi, _, _>(url, ()).await;
+
+    validator
+        .change_resource_control_policy(|policy| {
+            policy
+                .http_request_allow_list
+                .insert("localhost".to_owned());
+        })
+        .await;
+
+    chain
+        .graphql_mutation(application_id, "mutation { performHttpRequest }")
+        .await;
+
+    Ok(())
+}
+
+/// Tests if the contract rejects an invalid HTTP response sent by the service.
+#[test_log::test(tokio::test)]
+#[should_panic(expected = "Failed to execute block")]
+async fn contract_rejects_invalid_http_response_from_service() {
+    const HTTP_RESPONSE_BODY: &str = "Untrusted response";
+
+    let http_server =
+        HttpServer::start(Router::new().route("/", get(|| async { HTTP_RESPONSE_BODY })))
+            .await
+            .expect("Failed to start test HTTP server");
+    let port = http_server.port();
+    let url = format!("http://localhost:{port}/");
+
+    let (validator, application_id, chain) =
+        TestValidator::with_current_application::<Abi, _, _>(url, ()).await;
+
+    validator
+        .change_resource_control_policy(|policy| {
+            policy
+                .http_request_allow_list
+                .insert("localhost".to_owned());
+        })
+        .await;
+
+    chain
+        .graphql_mutation(application_id, "mutation { performHttpRequest }")
```

Comment:
It'd be nice to be able to really see that the block proposal failed for _exactly_ the expected reason here.